Improvements to `searx.data.CURRENCIES` initialization

return42 opened the pull request:
The PR is intended to ensure that the initialization of the searx.data.CURRENCIES does not unnecessarily extend the initialization time of SearXNG.

- https://github.com/searxng/searxng/issues/5223

BTW, the PR in the third commit contains a bugfix (for which I unfortunately have not found a test scenario).

Please read along the commit messages for more details.

return42 committed:
[mod] debug-log the bootload of the CURRENCIES caches (CurrenciesDB)

The time for the bootload is measured and recorded.  To have an eye on the
bootload time is motivated by: In [1] it can be seen that the bootload of the
CURRENCIES cache takes about 2/3 of the time required to initialize SearXNG::

    6.068 <module>  searx/webapp.py:1
    └─ 6.068 wrapper  pyinstrument/context_manager.py:52
       ├─ 5.538 init  searx/webapp.py:1373
       │  ├─ 4.822 initialize  searx/search/__init__.py:34
       │  │  ├─ 4.631 ProcessorMap.init  searx/search/processors/__init__.py:47
       │  │  │  ├─ 4.607 OnlineCurrencyProcessor.initialize  searx/search/processors/online_currency.py:55
       │  │  │  │  └─ 4.607 CurrenciesDB.init  searx/data/currencies.py:25
       │  │  │  │     ├─ 4.601 CurrenciesDB.load  searx/data/currencies.py:34
       │  │  │  │     │  ├─ 4.572 ExpireCacheSQLite.set  searx/cache.py:334

In the example, the CurrenciesDB.init call takes 4.6 seconds... on my laptop,
CurrenciesDB.init takes only 0.7 seconds.  The absolute numerical values depend
on external conditions, where I already find 4-5 seconds very long. Test::

    $ rm /tmp/sxng_cache_*.db*
    $ make run 2>&1 | grep "searx.data.CURRENCIES"
    DEBUG   searx.data : init searx.data.CURRENCIES
    DEBUG   searx.data : init searx.data.CURRENCIES added 9089 items in 0.7623255252838135 sec.

[1] https://github.com/searxng/searxng/issues/5223#issuecomment-3323083411

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] openstreetmap: fix CURRENCIES.iso4217_to_name

This patch is a leftover from PR-5204 [1].

[1] https://github.com/searxng/searxng/pull/5204

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] run bootload of the CURRENCIES cache on demand

In [1] it can be seen that the bootload of the CURRENCIES cache takes about 2/3
of the time required to initialize SearXNG.  Whatever the absolute durations may
be, an explicit bootload during the SearXNG initialization is not required, as
the bootload is already ensured by the API of the CACHE.

- ``CurrenciesDB.name_to_iso4217``
- ``CurrenciesDB.iso4217_to_name``

The fact that the bootload now occurs on-demand should improve the
initialization time of SearXNG.

[1] https://github.com/searxng/searxng/issues/5223#issuecomment-3323083411

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] ExpireCacheSQLite - implement .setmany() for bulk loading

[1] https://github.com/searxng/searxng/issues/5223#issuecomment-3328597147

Suggested-by: Ivan G <igabaldon@inetol.net> [1]
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

inetol commented on searx/data/tracker_patterns.py:
```diff
@@ -49,8 +49,20 @@ def init(self):
+            expire = None
+            rows.append((key, value, expire))
+
+        self.cache.setmany(rows, ctx=self.ctx_name)
```
```console
2025-09-25 09:17:11,751 ERROR:searx.plugins.tracker_url_remover: Exception while calling on_result
Traceback (most recent call last):
  File "/usr/local/searxng/searx/plugins/_core.py", line 270, in on_result
    ret = bool(plugin.on_result(request=request, search=search, result=result))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/tracker_url_remover.py", line 46, in on_result
    result.filter_urls(self.filter_url_field)
  File "/usr/local/searxng/searx/result_types/_base.py", line 567, in filter_urls
    _filter_urls(self, filter_func=filter_func)
  File "/usr/local/searxng/searx/result_types/_base.py", line 129, in _filter_urls
    new_url = filter_func(result, field_name, url_src)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/tracker_url_remover.py", line 58, in filter_url_field
    return TRACKER_PATTERNS.clean_url(url=url_src)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/data/tracker_patterns.py", line 121, in clean_url
    for rule in self.rules():
                ^^^^^^^^^^^^
  File "/usr/local/searxng/searx/data/tracker_patterns.py", line 81, in rules
    yield key, value[0], value[1]
                         ~~~~~^^^
IndexError: tuple index out of range
```

return42 replied:
> IndexError: tuple index out of range

is fixed

inetol reviewed (commented):
LGTM

return42 commented on searx/cache.py:
```diff
@@ -421,10 +421,14 @@ def _setmany(
             with self.connect() as conn:
+                conn.execute("BEGIN")
                 conn.executemany(sql_str, rows)
+                conn.commit()
```
> and instead of pushing all the statements in a single transaction, we still do it for each one"

@inetol:  that would be correct in a pure SQL session (auto-commit), but not when using the python API ..

[A `Connection` object can be used as a context manager that automatically commits or rolls back open transactions when leaving the body of the context manager.](https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager)

> So I wrapped #executemany() with `BEGIN; COMMIT;`:
>
> ```
> 0.065 CurrenciesDB.load  searx/data/currencies.py:35
> ```

Did you measured before (without BEGIN & commit) and after this commit? ..

inetol replied:
> @inetol: that would be correct in a pure SQL session (auto-commit), but not when using the python API ..
> 
> [A `Connection` object can be used as a context manager that automatically commits or rolls back open transactions when leaving the body of the context manager.](https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager)

The thing is we really need to control when to commit here, [autocommit](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.autocommit) was added in Python 3.12. The only way I found to avoid this behavior is to open a transaction manually, then it will not commit or rollback automatically.

> Did you measured before (without BEGIN & commit) and after this commit? ..

Profiling this block:
https://github.com/return42/searxng/blob/7da1201020eb9914ad75752becd5f0532b3503a3/searx/data/currencies.py#L39-L42

8f3fa2be5c290054dbf86e8d03b50e7732f80230
```console
Profile at /usr/local/searxng/searx/data/currencies.py:41

4.812 CurrenciesDB.load  searx/data/currencies.py:35
├─ 4.782 ExpireCacheSQLite.setmany  searx/cache.py:350
│  ├─ 4.780 ExpireCacheSQLite._setmany  searx/cache.py:375
│  │  ├─ 4.749 Connection.executemany  <built-in>
│  │  ├─ 0.015 ExpireCacheSQLite.create_table  searx/cache.py:284
│  │  │  ├─ 0.011 Connection.execute  <built-in>
│  │  │  ├─ 0.003 SQLiteProperties.set  searx/sqlitedb.py:386
│  │  │  │  └─ 0.003 Connection.execute  <built-in>
│  │  │  ├─ 0.001 ExpireCacheSQLite.connect  searx/sqlitedb.py:196
│  │  │  │  └─ 0.001 ExpireCacheSQLite._connect  searx/sqlitedb.py:190
│  │  │  │     └─ 0.001 Connection.execute  <built-in>
│  │  │  └─ 0.000 [self]  searx/cache.py
│  │  ├─ 0.009 [self]  searx/cache.py
│  │  ├─ 0.005 ExpireCacheSQLite.serialize  searx/cache.py:190
│  │  │  ├─ 0.004 dumps  <built-in>
│  │  │  └─ 0.001 [self]  searx/cache.py
│  │  ├─ 0.001 time  <built-in>
│  │  └─ 0.001 list.append  <built-in>
│  └─ 0.001 [self]  searx/cache.py
└─ 0.030 [self]  searx/data/currencies.py
```

7da1201020eb9914ad75752becd5f0532b3503a3
```console
Profile at /usr/local/searxng/searx/data/currencies.py:41

0.070 CurrenciesDB.load  searx/data/currencies.py:35
├─ 0.037 ExpireCacheSQLite.setmany  searx/cache.py:350
│  └─ 0.037 ExpireCacheSQLite._setmany  searx/cache.py:375
│     ├─ 0.014 Connection.executemany  <built-in>
│     ├─ 0.007 ExpireCacheSQLite.serialize  searx/cache.py:190
│     │  ├─ 0.005 dumps  <built-in>
│     │  └─ 0.002 [self]  searx/cache.py
│     ├─ 0.006 ExpireCacheSQLite.create_table  searx/cache.py:284
│     │  ├─ 0.004 Connection.execute  <built-in>
│     │  └─ 0.002 SQLiteProperties.set  searx/sqlitedb.py:386
│     │     └─ 0.002 Connection.execute  <built-in>
│     ├─ 0.004 Connection.commit  <built-in>
│     ├─ 0.004 [self]  searx/cache.py
│     └─ 0.002 list.append  <built-in>
└─ 0.033 [self]  searx/data/currencies.py
```

return42 replied:
Thanks for the detailed report.. very informative.. 

> [A `Connection` object can be used as a context manager that automatically commits or rolls back open transactions when leaving the body of the context manager.](https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager)

I have a suspicion why the `with` context manager does not start a transaction for us:

https://github.com/searxng/searxng/blob/87bc97776a5ba8ffa4bddf7eefcd3faeedddf3b1/searx/sqlitedb.py#L130

[If the connection attribute `isolation_level` is **not None**, new transactions are implicitly opened before execute() and executemany() executes INSERT, UPDATE, DELETE, or REPLACE statements;](https://docs.python.org/3/library/sqlite3.html#sqlite3-transaction-control-isolation-level)

I will have to investigate this a little more closely, since as of Python 3.12: [The recommended way of controlling transaction behaviour is through the `Connection.autocommit` attribute](https://docs.python.org/3/library/sqlite3.html#sqlite3-transaction-control-autocommit)

return42 committed:
[fix] sqlitedb: DB connector -- unset isolation_level

A Connection object can be used as a context manager that automatically commits
or rolls back open transactions when leaving the body of the context manager.

If the connection attribute ``isolation_level`` is **not** ``None``, new
transactions are implicitly opened before ``execute()`` and ``executemany()``
executes SQL- INSERT, UPDATE, DELETE, or REPLACE statements [1].

The old implementation set ``isolation_level`` to ``None`` and thereby prevented
the context manager from opening and closing the transaction scopes.

[1] https://docs.python.org/3/library/sqlite3.html#sqlite3-transaction-control-isolation-level
[2] https://github.com/searxng/searxng/pull/5239#discussion_r2381416731

Reported-by: Ivan G <igabaldon@inetol.net> [2]
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
